File: SDK/php/src/VerifioClient.php
<?php

namespace Verifio;

use GuzzleHttp\Client;
use GuzzleHttp\Exception\RequestException;
use GuzzleHttp\Exception\GuzzleException;
use Verifio\Exceptions\AuthenticationException;
use Verifio\Exceptions\InsufficientCreditsException;
use Verifio\Exceptions\NotFoundException;
use Verifio\Exceptions\RateLimitException;
use Verifio\Exceptions\ServerException;
use Verifio\Exceptions\ValidationException;
use Verifio\Exceptions\VerifioException;

class VerifioClient
{
    private string $apiKey;
    private string $baseUrl;
    private Client $httpClient;

    public function __construct(string $apiKey, ?string $baseUrl = null, ?Client $httpClient = null)
    {
        if (empty($apiKey)) {
            throw new ValidationException("API key is required");
        }

        $this->apiKey = $apiKey;
        $this->baseUrl = rtrim($baseUrl ?? 'https://verifio.email', '/');
        $this->httpClient = $httpClient ?? new Client([
            'timeout' => 30.0,
        ]);
    }

    /**
     * @param string $method
     * @param string $path
     * @param array|null $body
     * @return array|null Returns decoded JSON array or null
     * @throws VerifioException
     */
    public function request(string $method, string $path, ?array $body = null): ?array
    {
        $url = $this->baseUrl . "/api/verify/v1" . $path;

        $options = [
            'headers' => [
                'Authorization' => 'Bearer ' . $this->apiKey,
                'Content-Type'  => 'application/json',
                'Accept'        => 'application/json',
            ],
        ];

        if ($body !== null) {
            $options['json'] = $body;
        }

        try {
            $response = $this->httpClient->request($method, $url, $options);
            $statusCode = $response->getStatusCode();
            $content = $response->getBody()->getContents();

            $decoded = json_decode($content, true);

            if (json_last_error() !== JSON_ERROR_NONE) {
                throw new VerifioException("Invalid JSON response from server");
            }

            if ($statusCode >= 200 && $statusCode < 300) {
                if (isset($decoded['success']) && $decoded['success'] === true) {
                    return $decoded['data'] ?? null;
                }
                $this->handleError($statusCode, $decoded['error'] ?? null, $decoded['data'] ?? null);
            }

            $this->handleError($statusCode, null, null);

        } catch (RequestException $e) {
            $statusCode = 0;
            $errorMsg = $e->getMessage();
            $errorData = null;

            if ($e->hasResponse()) {
                $statusCode = $e->getResponse()->getStatusCode();
                $content = $e->getResponse()->getBody()->getContents();
                $decoded = json_decode($content, true);
                
                if (json_last_error() === JSON_ERROR_NONE) {
                    $errorMsg = $decoded['error'] ?? "HTTP Error " . $statusCode;
                    $errorData = $decoded['data'] ?? null;
                } else {
                    $errorMsg = "HTTP Error " . $statusCode;
                }
            }

            if ($statusCode > 0) {
                $this->handleError($statusCode, $errorMsg, $errorData);
            }

            throw new VerifioException("Network error: " . $errorMsg);
        } catch (GuzzleException $e) {
            throw new VerifioException("Client error: " . $e->getMessage());
        }
        
        return null;
    }

    /**
     * @throws VerifioException
     */
    private function handleError(int $status, ?string $message, ?array $data): void
    {
        $message = $message ?? "An error occurred";

        switch ($status) {
            case 401:
                throw new AuthenticationException($message);
            case 402:
                $remaining = $data['remaining'] ?? null;
                $required = $data['required'] ?? null;
                throw new InsufficientCreditsException($message, $remaining, $required);
            case 404:
                throw new NotFoundException($message);
            case 429:
                $retryAfter = $data['retry_after'] ?? null;
                throw new RateLimitException($message, $retryAfter);
            case 400:
                throw new ValidationException($message);
            case 500:
            case 502:
            case 503:
                throw new ServerException($message);
            default:
                throw new VerifioException($message, $status, null);
        }
    }
}

File: SDK/.net/VerifioClient.cs
using System;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using Verifio.Exceptions;
using Verifio.Models;

namespace Verifio
{
    public class VerifioConfig
    {
        public string ApiKey { get; set; } = string.Empty;
        public string? BaseUrl { get; set; }
    }

    public class VerifioClient
    {
        private readonly string _apiKey;
        private readonly string _baseUrl;
        private readonly HttpClient _httpClient;
        private readonly JsonSerializerOptions _jsonOptions;

        public VerifioClient(VerifioConfig config, HttpClient httpClient)
        {
            if (string.IsNullOrWhiteSpace(config.ApiKey))
            {
                throw new ValidationException("API key is required");
            }
            
            _apiKey = config.ApiKey;
            _baseUrl = (string.IsNullOrWhiteSpace(config.BaseUrl) ? "https://verifio.email" : config.BaseUrl).TrimEnd('/');
            _httpClient = httpClient;
            _httpClient.Timeout = TimeSpan.FromSeconds(30);

            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
                DefaultIgnoreCondition = System.Text.Json.Serialization.JsonIgnoreCondition.WhenWritingNull
            };
        }

        public async Task<T?> RequestAsync<T>(HttpMethod method, string path, object? body = null)
        {
            var url = $"{_baseUrl}/api/verify/v1{path}";
            var request = new HttpRequestMessage(method, url);
            
            request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);

            if (body != null)
            {
                var jsonBody = JsonSerializer.Serialize(body, _jsonOptions);
                request.Content = new StringContent(jsonBody, Encoding.UTF8, "application/json");
            }

            HttpResponseMessage response;
            try
            {
                response = await _httpClient.SendAsync(request);
            }
            catch (Exception ex)
            {
                throw new VerifioException($"Network error: {ex.Message}");
            }

            var contentString = await response.Content.ReadAsStringAsync();
            var status = (int)response.StatusCode;

            ApiResponse<T>? apiResponse = null;
            try
            {
                // We use a general JsonElement approach to cleanly parse errors too
                var rawRes = JsonSerializer.Deserialize<ApiResponse<JsonElement>>(contentString, _jsonOptions);
                if (rawRes != null)
                {
                    apiResponse = new ApiResponse<T>
                    {
                        Success = rawRes.Success,
                        Error = rawRes.Error,
                    };
                    
                    if (rawRes.Success && rawRes.Data.ValueKind != JsonValueKind.Undefined && rawRes.Data.ValueKind != JsonValueKind.Null)
                    {
                        apiResponse.Data = JsonSerializer.Deserialize<T>(rawRes.Data.GetRawText(), _jsonOptions);
                    }
                    
                    if (!rawRes.Success)
                    {
                        HandleError(status, rawRes.Error, rawRes.Data);
                    }
                }
            }
            catch (VerifioException) { throw; } // rethrow
            catch
            {
                // Parse failed
                if (status >= 400)
                {
                    HandleError(status, $"HTTP Error {status}", default);
                }
                throw new VerifioException("Invalid JSON response from server");
            }

            if (status < 200 || status >= 300 || (apiResponse is { Success: false }))
            {
                HandleError(status, apiResponse?.Error, default);
            }

            return apiResponse != null ? apiResponse.Data : default;
        }

        private void HandleError(int status, string? message, JsonElement data)
        {
            message ??= "An error occurred";

            int? GetIntField(string field)
            {
                if (data.ValueKind == JsonValueKind.Object && data.TryGetProperty(field, out var prop) && prop.TryGetInt32(out var val))
                    return val;
                return null;
            }

            switch (status)
            {
                case 401:
                    throw new AuthenticationException(message);
                case 402:
                    throw new InsufficientCreditsException(message, GetIntField("remaining"), GetIntField("required"));
                case 404:
                    throw new NotFoundException(message);
                case 429:
                    throw new RateLimitException(message, null);
                case 400:
                    throw new ValidationException(message);
                case 500:
                case 502:
                case 503:
                    throw new ServerException(message);
                default:
                    throw new VerifioException(message, status, null);
            }
        }
    }
}
